enhancements/cluster-logging/loki-storage.md

```diff
@@ -0,0 +1,628 @@
+
+### Goals
+
+The specific goals of this proposals are:
+
+* Outline the API to provide a declarative CRD for Loki deployments managed by the [loki-operator](https://github.com/viaq/loki-operator).
+
+* Propose API support into the ClusterLogging CRD for single entry point deployments.
+
+We will be successful when:
+
+* Users are able to declare a Loki instance via the CRD and the [loki-operator](https://github.com/viaq/loki-operator) rolls out the Loki deployment.
+
+* Users are able to declare Loki as alternative store in the Cluster Logging custom resource.
+
+* Users are able to operate Loki in Cluster Logging without the need to tune every single bit via the proposed CRD.
+
+### Non-Goals
+
+* We will not provide a general purpose API to install Loki at the field. The API is dedicated to cluster logging only.
+
+* We will not expose or implement all possible Loki configuration knobs in the API besides the most crucial to operate in cluster logging.
+
+* We will not expose Loki as a general log store for other ingestion sources than the ones supported and managed by cluster logging.
+
+* We will not offer support for custom labeling of logs other than the defaults provided by cluster logging collector.
+
+## Proposal
+
+To offer a seamless cluster logging experience with Loki as an alternative store, we propose to limit the API support to the following key features:
+
+* Deploy Loki **only** in microservices mode, i.e. each component is deployed and managed as an individual deployment/statefulset. This enables individual capacity planning and resource utilization per path ingestion/querying.
+
+* Multi-tenancy is enabled per default and requires the collector to apply the required [tenant identification](https://grafana.com/docs/loki/latest/operations/multi-tenancy/) on ingestion time. No user can disable this setting. In turn queries require the same identification information to be successful. Furthermore there is no support for querying across tenants.
+
+* Loki is operated only in a zero-dependency setup mode as described [here](https://grafana.com/docs/loki/latest/configuration/examples/#almost-zero-dependencies-setup). The only externally managed requirement is the availability of an S3-compatible API for object storage to persist logs.
+
+* Support a limited set of Loki scale outs, which are battle-tested on our internally managed Loki deployment. The target scale outs are defined using the format `Nx.{extra-small, small, medium}` where `N` the number of Loki deployments:
+  * **1x.extra-small**: Dedicated scale for development and demo purposes only! Not a valid configuration for cluster logging production environments.
+  * **1x.small**: Dedicated scale for low-retention log storage without redundancy support. All components are deployed with HA by running at least two replicas.
```

Comment:
It sound like a contradiction but in fact it is not. Redundancy does not refer to replicas here, but more to data processing redundancy. For example "single redundancy" means for Loki, that each log stream is processed on at least two ingesters. On the other hand you can still choose to have two ingester replicas but use "Zero Redundancy". From a traffic serving point you have two replicas always, so HA. However from a data processing point, you might loose data here, if a replica is killed and the handover is not finished to the other replica.

Reply:
I think we need to stop using the word "redundancy," because it's confusing. Loki doesn't mention the word "redundancy" in its code or documents that I can find. In Loki, Cortex, Thanos, etc they use "replication factor" which means how many replicas exist - which can be anything from one to infinity. When I hear "redundancy" I have to stop and think for a second. Can we please stick to the convention that other projects in our space use and avoid this confusion altogether?

Reply:
I have reworked Redundancy into ReplicationFactor. PTAL on the new wording.